docs/src/main/asciidoc/security-customization.adoc

```diff
@@ -620,6 +620,62 @@ are using an executor that is capable of propagating the identity (e.g. no `Comp
 to make sure that Quarkus can propagate it. For more information see the
 xref:context-propagation.adoc[Context Propagation Guide].
 
+[[observe-security-events]]
+== Observe security events
+
+Quarkus beans fire various security events that can be consumed by xref:cdi.adoc#events-and-observers[CDI observers].
+All security events implement the `io.quarkus.security.spi.runtime.SecurityEvent` interface, which allows you to consume all the events with one observer method.
+
+.List of supported security events
+
+* `io.quarkus.security.spi.runtime.AuthenticationFailureEvent`
+* `io.quarkus.security.spi.runtime.AuthenticationSuccessEvent`
+* `io.quarkus.security.spi.runtime.AuthorizationFailureEvent`
+* `io.quarkus.security.spi.runtime.AuthorizationSuccessEvent`
+* `io.quarkus.oidc.SecurityEvent`
+
+[[TIP]]
+For more information about security events specific to the Quarkus OpenID Connect extension, please see
+the xref:security-oidc-code-flow-authentication.adoc#listen-to-authentication-events[Listening to important authentication events]
+section of the OIDC code flow mechanism for protecting web applications guide.
+
+[source,java]
+----
+package org.acme.security;
+
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.security.spi.runtime.AuthenticationFailureEvent;
+import io.quarkus.security.spi.runtime.AuthorizationFailureEvent;
+import io.quarkus.security.spi.runtime.SecurityEvent;
+import io.vertx.ext.web.RoutingContext;
+import jakarta.enterprise.event.Observes;
+import jakarta.enterprise.event.ObservesAsync;
+
+public class SecurityEventObserver {
+
+    void observe(@ObservesAsync SecurityEvent event) {              <1>
```

Comment:
Should `SecurityEvent` have a category or type enum property, example, `AUTHENTICATION_FAILED` etc ? Otherwise one needs to do up to 5 `instanceof` checks to distinguish which event that is.  (Note OIDC SecurityEvent already has getType() returning its own enum, so it will be getEventType())

I wonder if it is really important to have the `SecurityEvent` interface ? Is it more effective to use it to capture all events ? I'm not sure users will treat all events generically, they will most likely need to know what type of event it is, in which case, they will observer `AuthenticationFailed` etc.

What do you think ?

Reply:
@michalvavrik 

>  Having categories and enum means anyone who want create new event needs to add his category or use very general category that makes it useless.

It is Quarkus which decides which types of events it can send so it will only be Quarkus which would add them

> How exactly are you going to map OIDC event type OIDC_FRONTCHANNEL_LOGOUT_COMPLETED to this new enum?

As I said, given that OIDC has its own event type, a property on this new `SecurityEvent` would have to be named differently. May be in case of OIDC specific events a new property would be set to `OIDC` and then OIDC specific type would be used to narrow down further.

But, in any case, it can be added later if necessary.

But, IMHO, we should not recommend users writing the code like the one you suggest above. It still does not give any hint what this event is about. I totally agree some users may want to do fine-grained checks, check class names, individual properties, but in general, IMHO, we should advise a different pattern, where users catch specific events, then log them and then use `SecurityEvent` interface to post to Kafka or log properties - this will support both scenarios - where users are expressing interest in specific events but also using a common interface for further processing. IMHO it would be better than only giving an example where all events can be captured by a single observer.

Please consider updating the example if it sounds reasonable

Reply:
I'm fine with dropping example of `void observe(@ObservesAsync SecurityEvent event) {`.


I don't think you can deny that:

- many application won't want to log something specific to the event (like _Authentication successful ...._) because extensive logging may slows down app, that's way they may not want to deal with anything specific to the event type
- if oidc `SecurityEvent` put `io.quarkus.oidc.SecurityEvent.Type` also to `eventProperties` map, than I don't loose any information by using interface as I do have
  - `SecurityIdentity`
  - `eventProperties`
  - `event class` which tells my type of event
-  if you use observer specific for every event like OIDC SecurityEvent, you will still need some identifier when you are storing it, I'm not sure why class signature is any worse than something you can type

My conclusion is:

1. I'll drop `SecurityEvent` observer example in favor of all other event type observers
2. I'll keep the interface.

If you disagree with 2. please comment.

Reply:
@michalvavrik Right, your points about keeping the interface are convincing, I'm just suggesting to revise the example for possibly a more typical case where specific even types are observed but also, the fact that all of them implement the interface can be used for some common event processing for all of those types like logging of the properties, posting to Kafka etc